storage/innobase/dict/dict0crea.cc

```diff
@@ -1843,3 +1970,388 @@ dict_create_add_tablespace_to_dictionary(
+		"  SELECT ID FROM SYS_ZIP_DICT\n"
+		"    WHERE NAME = :dict_name;\n"
+		"BEGIN\n"
+		"  found := 1;\n"
+		"  OPEN cur;\n"
+		"  FETCH cur INTO my_func();\n"
+		"  IF SQL % NOTFOUND THEN\n"
+		"    found := 0;\n"
+		"  END IF;\n"
+		"  CLOSE cur;\n"
+		"  IF found = 1 THEN\n"
+		"    INSERT INTO SYS_ZIP_DICT_COLS VALUES"
+		"      (:table_id, :column_pos, :dict_id);\n"
+		"  END IF;\n"
+		"END;\n",
+		FALSE, trx);
+	if(error == DB_SUCCESS)
+	{
+		ib_uint32_t dict_id = mach_read_from_4(reinterpret_cast<const byte*>(&dict_id_buf));
+		if(dict_id == ULINT32_UNDEFINED)
+			error = DB_RECORD_NOT_FOUND;
+	}
+	return error;
+}
+
+/********************************************************************//**
+Get a single compression dictionary id for the given (table id, column pos)
+pair.
+@return	error code or DB_SUCCESS */
+UNIV_INTERN
+dberr_t
+dict_create_get_zip_dict_id_by_reference(
+/*=====================================*/
+	ulint	table_id,	/*!< in: table id */
+	ulint	column_pos,	/*!< in: column position*/
+	ulint*	dict_id,	/*!< out: column position*/
+	trx_t*	trx)		/*!< in/out: transaction */
+{
+	ut_ad(dict_id);
+
+	pars_info_t* info = pars_info_create();
+
+	ib_uint32_t dict_id_buf;
+	mach_write_to_4(reinterpret_cast<byte*>(&dict_id_buf ), ULINT32_UNDEFINED);
+
+	pars_info_add_int4_literal(info, "table_id", table_id);
+	pars_info_add_int4_literal(info, "column_pos", column_pos);
+	pars_info_bind_function(
+		info, "my_func", dict_create_extract_int_aux, &dict_id_buf);
+
+	dberr_t error = que_eval_sql(info,
+		"PROCEDURE P () IS\n"
+		"DECLARE FUNCTION my_func;\n"
+		"DECLARE CURSOR cur IS\n"
+		"  SELECT DICT_ID FROM SYS_ZIP_DICT_COLS\n"
+		"    WHERE TABLE_ID = :table_id AND\n"
+		"          COLUMN_POS = :column_pos;\n"
+		"BEGIN\n"
+		"  OPEN cur;\n"
+		"  FETCH cur INTO my_func();\n"
+		"  CLOSE cur;\n"
+		"END;\n",
+		FALSE, trx);
+	if(error == DB_SUCCESS)
+	{
+		ib_uint32_t local_dict_id = mach_read_from_4(reinterpret_cast<const byte*>(&dict_id_buf));
+		if(local_dict_id == ULINT32_UNDEFINED)
+			error = DB_RECORD_NOT_FOUND;
+		else
+			*dict_id = local_dict_id;
+	}
+	return error;
+}
+
+/******************************************************************//**
+Fetch callback, just stores extracted zip_dict data in the external
+variable.
+@return always returns TRUE */
+static
+ibool
+dict_create_get_zip_dict_info_by_id_aux(
+/*===================*/
+	void*	row,		/*!< in: sel_node_t* */
+	void*	user_arg)	/*!< in: pointer to ib_vector_t */
+{
+	sel_node_t*	node = static_cast<sel_node_t*>(row);
+	LEX_STRING* tuple = static_cast<LEX_STRING*>(user_arg);
+
+	/* NAME field */
+	que_node_t*	exp = node->select_list;
+	ut_a(exp != 0);
+
+	dfield_t*	dfield = que_node_get_val(exp);
+	dtype_t*	type = dfield_get_type(dfield);
+	ut_a(dtype_get_mtype(type) == DATA_VARCHAR);
+
+	ulint		len = dfield_get_len(dfield);
+	void*		data = dfield_get_data(dfield);
+
+
+	if (len != UNIV_SQL_NULL) {
+		tuple[0].str = (char*)my_malloc(len + 1, MYF(0));
```

Comment:
`dict_create_get_zip_dict_info_by_id_aux()` / `dict_create_get_zip_dict_info_by_id()` reworked. They now use a new struct `zip_dict_info_aux` to return zip dict name, zip dict data and status code instead of `LEX_STRING tuple[2]`. OOM error is reported via this status code.